# Biên tập dữ liệu

In [3]:
setwd('E:/Statistics-R/THONG KE R')
chol <- read.table('chol.txt', header=TRUE)
attach(chol)

 ## 4.1 Kiểm tra số liệu trống không (missing value)
Trong nghiên cứu, vì nhiều lí do số liệu không thể thu thập được cho tất cả đối
tượng, hay không thể đo lường tất cả biến số cho một đối tượng. Trong trường hợp đó,
số liệu trống được xem là “missing value” (mà tôi tạm dịch là số liệu trống không). R
xem các số liệu trống không là NA. Có một số kiểm định thống kê đòi hỏi các số liệu
trống không phải được loại ra (vì không thể tính toán được) trước khi phân tích. R có
một lệnh rất có ích cho việc này: na.omit, và cách sử dụng như sau:

In [4]:
chol.new <- na.omit(chol)

## 4.2 Tách rời dữ liệu: subset
Nếu chúng ta, vì một lí do nào đó, chỉ muốn phân tích riêng cho nam giới, chúng
ta có thể tách chol ra thành hai data.frame, tạm gọi là nam và nu. Để làm chuyện này,
chúng ta dùng lệnh subset(data, cond), trong đó data là data.frame mà chúng ta
muốn tách rời, và cond là điều kiện. Ví dụ:

In [6]:
nam <- subset(chol, sex=='Nam')
nu <- subset(chol, sex=='Nu')

In [7]:
old <- subset(chol, age>=60)
dim(old)

[1] 25  8

In [9]:
n60 <- subset(chol, age>=60 & sex=='Nam')
dim(n60)

[1] 10  8

### 4.3 Chiết số liệu từ một data .frame

In [10]:
data2 <- chol[, c(1,3,7)]

In [11]:
data3 <- chol[1:10, c(1,3,7)]
print(data3)

   id age  tc
1   1  57 4.0
2   2  64 3.5
3   3  60 4.7
4   4  65 7.7
5   5  47 5.0
6   6  65 4.2
7   7  76 5.9
8   8  61 6.1
9   9  59 5.9
10 10  57 4.0


### 4.5 Mã hóa số liệu (data coding)
Trong việc xử lí số liệu dịch tễ học, nhiều khi chúng ta cần phải biến đổi số liệu từ biến
liên tục sang biến mang tính cách phân loại. Chẳng hạn như trong chẩn đoán loãng
xương, những phụ nữ có chỉ số T của mật độ chất khoáng trong xương (bone mineral
density hay BMD) bằng hay thấp hơn -2.5 được xem là “loãng xương”, những ai có
BMD giữa -2.5 và -1.0 là “xốp xương” (osteopenia), và trên -1.0 là “bình thường”. Ví
dụ, chúng ta có số liệu BMD từ 10 bệnh nhân như sau:

In [13]:
bmd <- c(-0.92,0.21,0.17,-3.21,-1.80,-2.60,-2.00,1.71,2.12,-2.11)

Để phân loại 3 nhóm loãng xương, xốp xương, và bình thường, chúng ta có thể dùng mã
số 1, 2 và 3. Nói cách khác, chúng ta muốn tạo nên một biến số khác (hãy gọi là
diagnosis) gồm 3 giá trị trên dựa vào giá trị của bmd. Để làm việc này, chúng ta sử
dụng lệnh:

In [14]:
# tạm thời cho biến số diagnosis bằng bmd
diagnosis <- bmd
# biến đổi bmd thành diagnosis
diagnosis[bmd <= -2.5] <- 1
diagnosis[bmd > -2.5 & bmd <= 1.0] <- 2
diagnosis[bmd > -1.0] <- 3
# tạo thành một data frame
data <- data.frame(bmd, diagnosis)

In [16]:
# liệt kê để kiểm tra xem lệnh có hiệu quả không
data

bmd   diagnosis
1  -0.92 3        
2   0.21 3        
3   0.17 3        
4  -3.21 1        
5  -1.80 2        
6  -2.60 1        
7  -2.00 2        
8   1.71 3        
9   2.12 3        
10 -2.11 2

### 4.5.1 Biến đổi số liệu bằng cách dùng replace
Một cách biến đổi số liệu khác là dùng replace, dù cách này có vẻ rườm rà chút ít.
Tiếp tục ví dụ trên, chúng ta biến đổi từ bmd sang diagnosis như sau:

In [17]:
diagnosis <- bmd
diagnosis <- replace(diagnosis, bmd <= -2.5, 1)
diagnosis <- replace(diagnosis, bmd > -2.5 & bmd <= 1.0, 2)
diagnosis <- replace(diagnosis, bmd > -1.0, 3)

### 4.5.2 Biến đổi thành yếu tố (factor)
Trong phân tích thống kê, chúng ta phân biệt một biến số mang tính yếu tố (factor) và
biến số liên tục bình thường. Biến số yếu tố không thể dùng để tính toán như cộng trừ
nhân chia, nhưng biến số số học có thể sử dụng để tính toán. Chẳng hạn như trong ví dụ
bmd và diagnosis trên, diagnosis là yếu tố vì giá trị trung bình giữa 1 và 2 chẳng
có ý nghĩa thực tế gì cả; còn bmd là biến số số học.

Nhưng hiện nay, diagnosis được xem là một biến số số học. Để biến thành biến số
yếu tố, chúng ta cần sử dụng function factor như sau:


In [18]:
diag <- factor(diagnosis)
diag

[1] 3 3 3 1 2 1 2 3 3 2
Levels: 1 2 3

Chú ý R bây giờ thông báo cho chúng ta biết diag có 3 bậc: 1, 2 và 3. Nếu chúng ta yêu
cầu R tính số trung bình của diag, R sẽ không làm theo yêu cầu này, vì đó không phải là
một biến số số học:

In [19]:
mean(diag)

Warning message in mean.default(diag):
"argument is not numeric or logical: returning NA"


[1] NA

Dĩ nhiên, chúng ta có thể tính giá trị trung bình của diagnosis:

In [20]:
mean(diagnosis)

[1] 2.3

nhưng kết quả 2.3 này không có ý nghĩa gì trong thực tế cả.

### 4.6 Chia nhóm bằng cut

Với một biến liên tục, chúng ta có thể chia thành nhiều nhóm bằng hàm cut. Ví dụ,
chúng ta có biến age như sau:

In [21]:
age <- c(17,19,22,43,14,8,12,19,20,51,8,12,27,31,44)

Độ tuổi thấp nhất là 8 và cao nhất là 51. Nếu chúng ta muốn chia thành 2 nhóm tuổi:

In [22]:
cut(age, 2)

[1] (7.96,29.5] (7.96,29.5] (7.96,29.5] (29.5,51]   (7.96,29.5] (7.96,29.5]
 [7] (7.96,29.5] (7.96,29.5] (7.96,29.5] (29.5,51]   (7.96,29.5] (7.96,29.5]
[13] (7.96,29.5] (29.5,51]   (29.5,51]  
Levels: (7.96,29.5] (29.5,51]

cut chia biến age thành 2 nhóm: nhóm 1 tuổi từ 7.96 đến 29.5; nhóm 2 từ 29.5 đến
51. Chúng ta có thể đếm số đối tượng trong từng nhóm tuổi bằng hàm table như sau:

In [23]:
table(cut(age, 2))


(7.96,29.5]   (29.5,51] 
         11           4 

In [24]:
ageg <- cut(age, 3, labels=c("low", "medium", "high"))

In [25]:
ageg <- cut(age, 3, labels=c("low", "medium", "high"))
table(ageg)

ageg
   low medium   high 
    10      2      3 

Tất nhiên, chúng ta cũng có thể chia age thành 4 nhóm (quartiles) bằng cách cho những
thông số 0, 0.25, 0.50 và 0.75 như sau:

In [33]:
cut(age, 
    breaks=quantile(age, c(0, 0.25, 0.50, 0.75, 1)), 
    labels=c('q1', 'q2', 'q3', 'q4'),
    include.lowest=TRUE)

[1] q2 q2 q3 q4 q2 q1 q1 q2 q3 q4 q1 q1 q3 q4 q4
Levels: q1 q2 q3 q4

In [36]:
cut(age,
    breaks=quantile(c(0, 0.25, 0.50, 0.75, 1)),
    labels=c('q1', 'q2', 'q3', 'q4'),
    include.lowest=TRUE)

[1] <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA> <NA>
Levels: q1 q2 q3 q4

### Tham khảo thêm hàm quantile

In [31]:
#creates a vector having some values and the quantile function will return the percentiles for the data.

df<-c(12,3,4,56,78,18,46,78,100)
quantile(df)

0%  25%  50%  75% 100% 
   3   12   46   78  100

### 4.7. Tập hợp số liệu bằng cut2 (Hmisc)

Hàm cut trên chia biến số theo giá trị của biến, chứ không dựa vào số mẫu, cho
nên số lượng mẫu trong từng nhóm không bằng nhau. Tuy nhiên, trong phân tích thống
kê, có khi chúng ta cần phải phân chia một biến số liên tục thành nhiều nhóm dựa vào
phân phối của biến số nhưng số mẫu bằng hay tương đương nhau. Chẳng hạn như đối
với biến số bmd chúng ta có thể “cắt” dãy số thành 3 nhóm với số mẫu tương đương nhau
bằng cách dùng function cut2 (trong thư viện Hmisc) như sau:

In [2]:
# nhập thư viện Hmisc để có thể dùng function cut2
library(Hmisc)
bmd <- c(-0.92,0.21,0.17,-3.21,-1.80,-2.60,-2.00,1.71,2.12,-2.11)
# chia biến số bmd thành 2 nhóm và để trong đối tượng group
group <- cut2(bmd, g=2)
table(group)


Attaching package: 'Hmisc'


The following objects are masked from 'package:base':

    format.pval, units




group
[-3.21,-0.92) [-0.92, 2.12] 
            5             5 

Như thấy qua ví dụ trên, g = 2 có nghĩa là chia thành 2 nhóm (g = group). R tự động
chia thành nhóm 1 gồm giá trị bmd từ -3.21 đến -0.92, và nhóm 2 từ -0.92 đến 2.12. Mỗi
nhóm gồm có 5 số.

Tất nhiên, chúng ta cũng có thể chia thành 3 nhóm bằng lệnh:

In [3]:
group <- cut2(bmd, g=3)

Và với lệnh table chúng ta sẽ biết có 3 nhóm, nhóm 1 gồm 4 số, nhóm 2 và 3 mỗi nhóm
có 3 số:

In [4]:
table(group)

group
[-3.21,-1.80) [-1.80, 0.21) [ 0.21, 2.12] 
            4             3             3 